### PR TITLE
`usesResource` in `ConditionDBWriter`

### PR DESCRIPTION
**Pull request by mmusich**
#### PR description:

This is a follow-up to https://github.com/cms-sw/cmssw/pull/40754#issuecomment-1431675916: 
   * introduce `cond::service::PoolDBOutputService::kSharedResource`
   * use the `cond::service::PoolDBOutputService::kSharedResource` resource in `ConditionDBWriter<T>`
   * rationalize `usesResource()` in the classes in all classes in the `ConditionDBWriter<T>` hierarchy and remove unnecessary dependencies

#### PR validation:

`cmssw` compiles.

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

N/A

### CODE REVIEW

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by emanueleusai** [comment]
+1

**Comment by francescobrivio** [comment]
assign core

Changes are in line with PR description and the discussion from which this PR originated (https://github.com/cms-sw/cmssw/pull/40754#issuecomment-1431675916), but I'd like to hear @makortel opinion if this is what he had in mind from the fwk point of view.

**Comment by tvami** [comment]
+1
 * tests pass

File: CalibTracker/SiStripChannelGain/plugins/DeDxDiscriminatorLearner.cc [hunk 1]
```diff
@@ -33,6 +33,7 @@ using namespace edm;
 
 DeDxDiscriminatorLearner::DeDxDiscriminatorLearner(const edm::ParameterSet& iConfig)
     : ConditionDBWriter<PhysicsTools::Calibration::HistogramD3D>(iConfig) {
+  usesResource();
```
Comment (makortel) [inline CalibTracker/SiStripChannelGain/plugins/DeDxDiscriminatorLearner.cc:36]: I see now that my comment https://github.com/cms-sw/cmssw/pull/40754#issuecomment-1431675916 was confusing, apologies. I should have written that, especially now that the bcase class `CondDBWriter<T>` calls the `usesResource(cond::service::PoolDBOutputService::kSharedResource)`, the deriving classes do not need to call `usesResource()`, unless they use some additional shared resource. I would naively expect that most of the deriving classes do not need the call at all, and the ones that use `TFileService` would need `usesResource(TFileService::kSharedResource)`.

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by tvami** [comment]
+1
 * tests pass
 * PR according to the desc and the suggestions by Matti

**Comment by makortel** [comment]
+core

**Comment by perrotta** [comment]
+1

**Pull request merged**